### PR TITLE
WFLY-4956 Automatically enable Hibernate Search in deployments

### PR DESCRIPTION
**Pull request by scottmarlow**

### CODE REVIEW

**Comment by ctomc** [comment]
Is this really the best place to have this? 
Given that hibernate search is optional module and will probably come in via feature pack, shouldn't this be part of said feature pack and not be part of JPA code/subsystem.

**Comment by Sanne** [comment]
Looks great, thanks for making this!

**Comment by Sanne** [comment]
@ctomc Hibernate Search is not an optional module. But let's assume we want to include it in our feature pack, that would have the Search project depend on WildFly, and create a circular dependency.
Or I'm misunderstanding? I still have to catch up on the feature packs.

**Comment by scottmarlow** [comment]
Last I heard, Hibernate Search is staying in WildFly.  If it moves into a feature pack, Search needs its own subsystem that is part of the feature pack.
From #hibernate-dev irc conversation about this change:
Jul 16 11:32:35 <smarlow>       sannegrinovero: as long as you don't remove Search from WildFly
Jul 16 11:34:56 <sannegrinovero>        smarlow: no we won't remove Search from Wildfly

Perhaps I misunderstood what is meant by keeping Search in WildFly.  If Search might move into a feature pack, then the JPA subsystem/deployer should not contain this change.  The disadvantage of moving Search to a feature pack, is that you wouldn't be able to depend on persistence unit properties (you could keep the annotation scanning though).

**Comment by ctomc** [comment]
No hsearch wont be removed from WildFly but it will be included in WildFly by feature pack. 
Which the end results in same wildfly distribution.
But my concern is that JPA subsystem already does too much for what it is meant. It handles JPA deployments, hibernate deployments, different JPA implementations and now also hibernate search.

That is why I am saying this should be probably be separated into its own subsystem.

When it comes to splitting wildfly into more fine grained feature packs, this big integration peace makes it harder to split in to smaller feature packs.
For example with change like this PR it would require that if we ever do JPA feature pack, that said feature pack would also need to depend not only to all jpa & hibernate related feature packs / modules but also hibernate search. Which is by all means an optional addon. With having integration on "parent" dependency you cannot really make it proper optional addon.

**Comment by Sanne** [comment]
I see your point @ctomc but I'm going to need to talk to you on two things:
 - do you expect the Hibernate projects to depend on Wildfly - to include such deployers?
 - who's going to define what the version is for module slot "main" ?

By depending on WildFly we introduce circular dependencies, and honestly also a level of expertise about how the containers work which we can't really afford to maintain. Say for example you change some APIs of the deployer, the "catch up" cycle is going to be a nightmare.

For Hibernate users it's important to be able to run multiple different versions on the container. I get this question very often, but we don't seem to be working to make that possible - even though the modules system would allow this so nicely.

To me that means that we should not ship a feature pack as "main" slot but using the specific version, 
so that people can overlay multiple versions, and the integration code *part of WildFly* is in charge of choosing a default one to use as "main" (and to have injected by the above deployer by default".

File: jpa/subsystem/src/main/java/org/jboss/as/jpa/messages/JpaLogger.java [hunk 1]
```diff
@@ -753,4 +753,6 @@
     @Message(id = 70, value = "A container-managed extended persistence context can only be initiated within the scope of a stateful session bean (persistence unit '%s').")
     IllegalStateException xpcOnlyFromSFSB(String scopedPuName);
 
+    @Message(id = 71, value = "deployment '%s' specified more than one Hibernate Search module name ('%s','%s')")
```
Comment (jamezp) [inline jpa/subsystem/src/main/java/org/jboss/as/jpa/messages/JpaLogger.java]: Not to be nit-picky, but this should start with `Deployment` (capital D).

**Comment by n1hility** [comment]
The only reason to use a a feature-pack is if you are building a custom distribution of wildfly. If you are not building a custom distribution of WildFly then it does not make sense to define a feature pack. Notably if you are producing an add-on distribution, this would not involve a feature pack. Also it should be mentioned that we do not allow subsystems that are part of the official WildFly distributions to be pulled in from maven deps. They must be part of the WF codebase for both auditing and compatibility reasons.

**Comment by jamezp** [comment]
The only question I have on this is do we want Hibernate Search to be a dependency on all JPA deployments? I don't know a lot about how what we do in this area, but IIRC a user can define a different JPA provider. Will this also add the Hibernate Search module dependency to those deployments?

**Comment by ctomc** [comment]
Maybe we could have hibernate-search subsystem that would do just this? 
As jpa subsystem does feel bit bloated with all the stuff it does.
@jamezp you make a good point here, in case user us using non hibernate JPA impl and uses hibernate search annotations it will add hibernate search deps which could cause it to fail deployment.
Sure it is a user deployment error but still.

**Comment by Sanne** [comment]
@n1hility 

    The only reason to use a a feature-pack is if you are building a custom distribution of wildfly.

That would be a nice clarification but I'm surprised as it contradicts suggestions I gathered from others.
Wasn't there also a "call to actions" to have all libraries moved into feature packs to help wildfly-swarm use them?

From my perspective, what we're primarily after is to not have to keep the modules in wildfly in-synch with the ones we build within Hibernate Search, as there's overlap in testing needs. But fetching the latest modules and "sedding" the XML definitions to replace the slot identifier would be good enough for this...


@jamezp good point, in fact this was on my original spec I shared with scott but we forgot about it.
It gets a bit complex as people can also have multiple JPA providers. BTW as far as I know, Hibernate ORM is currently always added to the classpath even when using a different provider (had quite some issues for Hibernate OGM with that: we're going to follow up with improvements in that area)
So I don't think it would fail the deployment by having Hibernate Search around as well - but would be a good improvement to avoid that.

**Comment by n1hility** [comment]
That would be a nice clarification but I'm surprised as it contradicts suggestions I gathered from others.
    Wasn't there also a "call to actions" to have all libraries moved into feature packs to help wildfly-swarm     use them?

**Comment by Sanne** [comment]
To clarify: Hibernate Search won't "do" anything unless it's loaded by Hibernate ORM. The only downside of having it unintentionally on classpath is classpath bloat and possible conflicts with other libraries.

**Comment by n1hility** [comment]
That would be a nice clarification but I'm surprised as it contradicts suggestions I    
    gathered from others.

I apologize for the confusion generated. I posted what I hope should clarify the issue to the ml here: 

http://lists.jboss.org/pipermail/wildfly-dev/2015-August/004278.html


     Wasn't there also a "call to actions" to have all libraries moved into feature packs
      to help wildfly-swarm use them?

Not that I am aware of. Swarm can use anything thats in the WildFly distro


    From my perspective, what we're primarily after is to not have to keep the modules    
    in wildfly in-synch with the ones we build within Hibernate Search, as there's 
    overlap in testing needs. But fetching the latest modules and "sedding" the XML 
    definitions to replace the slot identifier would be good enough for this...

I’m not sure how that can be achieved as WildFly has a pom that locks down versions of everything, including transitives, so no matter what we need an update PR. The slots for everything in wildfly should just be the default “master” unless we have a good justification for shipping two of something and then a different slot starts to make sense.

**Comment by Sanne** [comment]
All, this discussion continued on wildfly's hipchat channel. Please see the log for details but the conclusion is that this code belongs in WildFly.

+1 to merge here a (polished) version of this patch.

Regarding the issue of injecting Hibernate Search for non-Hibernate deployments: as I mentioned above failing to improve on that wouldn't cause any immediate failure, but I'd also prefer to see that improved.

Specifically we wouldn't just need to check that Hibernate (ORM) is being used as JPA provider for (at least one) of the Persistence Units, but we should also check that the "main" slot of Hibernate ORM is being used exclusively.

The "main" slot of Hibernate Search is designed to be compatible with the "main" slot of Hibernate ORM. Other combinations should be possible too but only if the user is setting the modules explicitly.

**Comment by scottmarlow** [comment]
@jamezp "Will this also add the Hibernate Search module dependency to those deployments?" No, we do not add Hibernate Search to JPA applications by default.  This is about trying to detect if the application is using Hibernate Search and adding it if yes.  This code could be moved to a Hibernate Search subsystem if we ever add one.  Also, thanks for the capitalization comment!

**Comment by scottmarlow** [comment]
@sanne, currently we check if any persistence unit contains "wildfly.jpa.hibernate.search.module" or the application has any "org.hibernate.search.annotations.Indexed" annotations.  Regarding adding a check to see if the application is using any non-Hibernate ORM persistence providers, I'm not sure about that yet.  

Regarding how we are always adding ORM jars for OGM, we just need a unit test that reproduces that (shouldn't be a big deal to address).  I think that I will be able to reproduce it without using OGM.  We have been adding other persistence providers fine for 4 years now (I think the bug is limited to certain conditions that OGM sadly experiences).

**Comment by Sanne** [comment]
Ok. It leaves you open to people using Hibernate Search annotations but not intentioned to use it.. but I agree it's far fetched, and they still can disable it explicitly if they really want to, and it's harmless.

The only non-harmless scenario I can think of is injecting the main Hibernate Search module when the user *is* using Hibernate but setting a different Hibernate ORM slot (or have it application provided). In this case if they don't also disable your deployer they will get into trouble.

Looks like that the more polished version would require high coupling with the JPA deployer? Or at least clearly document the need to set the Hibernate Search module property as well, when setting the Hibernate ORM module property.

**Comment by scottmarlow** [comment]
@ctomc 
>@jamezp you make a good point here, in case user us using non hibernate JPA impl and uses hibernate search annotations it will add hibernate search deps which could cause it to fail deployment.
Sure it is a user deployment error but still.

Using Hibernate Search with non-Hibernate ORM persistence providers (e.g. EclipseLink/OpenJPA/others) will fail deployment I think (likely with a CNFE).  I'm not really seeing the advantage yet of filtering against applications that containing zero Hibernate ORM based persistence providers, since those applications will fail to deploy with a CNFE.

**Comment by Sanne** [comment]
> Using Hibernate Search with non-Hibernate ORM persistence providers (e.g. EclipseLink/OpenJPA/others) will fail deployment I think (likely with a CNFE). I'm not really seeing the advantage yet of filtering against applications that containing zero Hibernate ORM based persistence providers, since those applications will fail to deploy with a CNFE.

No, no error will be throw in such case. As I mentioned in previous post, the more concerning combination is to have Hibernate Search automatically enabled when an incompatible version of Hibernate ORM is being used.

**Comment by scottmarlow** [comment]
> No, no error will be throw in such case. As I mentioned in previous post, the more concerning combination is to have Hibernate Search automatically enabled when an incompatible version of Hibernate ORM is being used.

No CNFE error when application entity class definitions are being loaded and Hibernate Search cannot be found?

**Comment by ctomc** [comment]
Well, it is fine if deployment fails, as long as user gets some reasonable error to let him know that he cannot mix non hibernate JPA provider with hibernate search.
Anyhow I think we shouldn't complicate too much at this point. It would be just worth doing some tests how it behaves in different scenarios.

**Comment by jamezp** [comment]
@scottmarlow Sorry I missed this line :) https://github.com/wildfly/wildfly/pull/7862/files#diff-d0d7b17c8881f552a96a4a17a2396eb9R110.

**Comment by scottmarlow** [comment]
Thanks @jamezp, capitalized the Logger message.

**Comment by n1hility** [comment]
Yesterday in a discussions with @Sanne, he mentioned that it was very important that users use module imports to include a different version of hibernate search. It looks like this DUP would end up conflicting with that, there would be two imports on search.

**Comment by scottmarlow** [comment]
That will happen if the application doesn't specify "wildfly.jpa.hibernate.search.module=none" in a persistence unit.

**Comment by n1hility** [comment]
@scottmarlow so is the intention that they specify the exact hibernate search module they wanted in persistence.xml?

**Comment by scottmarlow** [comment]
@n1hility yes, the application can specify the exact Hibernate Search module (e.g. wildfly.jpa.hibernate.search.module=mysearchmodule:xyz) that they want in one of the persistence unit definitions (we fail the deployment if they specify more than one Hibernate Search module).  Mostly, applications will not specify the Hibernate Search module in the persistence.xml, instead they will use the "org.hibernate.search.orm:main".  If the application doesn't specify the "wildfly.jpa.hibernate.search.module" in a persistence unit, we check if any application class has a "org.hibernate.search.annotations.Indexed" and add a dependency on "org.hibernate.search.orm:main".  

Currently, applications can specify the Hibernate Search module via MANIFEST.MF/jboss-deployment-structure.xml, however, there is currently no automatic Hibernate Search dependencies added based on applications using Hibernate Search dependencies.  

Just thinking some more about the issue that you previously mentioned, I wonder how applications will know to stop using the MANIFEST.MF/jboss-deployment-structure.xml for specifying the Hibernate Search module name.

**Comment by Sanne** [comment]
@n1hility True, *currently* users have to explicitly depend on the module.

The intention here is to simplify that - as long as it's a safe guess and the property allows an override.

Average user should get it to work by simply adding the right annotation, w/o need for a custom deployment structure definition. Power user should be able to override it, and the proposed attribute allows for that.

**Comment by n1hility** [comment]
@Sanne @scottmarlow another question is how would this work if hibernate search was an add-on and not shipped in the distro, what would the "default" be. The master slot?

**Comment by Sanne** [comment]
@n1hility 
 > would this work if hibernate search was an add-on and not shipped in the distro, what would the "default" be. The master slot?

I think master would be ok, especially as we would be using the aliases strategy. But in that case wouldn't this same deployer be part of the add-on?

**Comment by Sanne** [comment]
the gist is this is +1 to merge right?
(Good for me)

**Comment by scottmarlow** [comment]
If Hibernate Search is changed to an add-on, we would need to delete references to Hibernate Search from the JPA deployer.

**Comment by scottmarlow** [comment]
Just to qualify my comment with a little more detail, the JPA subsystem does not explicitly depend on any Hibernate Search module classes but instead looks for Hibernate Search annotations by class name in string form.  If Hibernate Search was changed to be an add-on, we would need to update the JPA subsystem to not automatically add Hibernate Search dependencies to deployments.

**Comment by scottmarlow** [comment]
Sounds to me like we need to discuss again whether Hibernate Search could ever become an add-on, if yes, Hibernate Search would need its own deployer.  I think that is the concern about the JPA subsystem being coupled to Hibernate Search.

**Comment by Sanne** [comment]
I don't understand what else there is to discuss but please let's chat about it and try to deliver it quickly as we've been planning this little improvement since a year?

**Comment by Sanne** [comment]
Discussed this over chat again with @scottmarlow and @n1hility and agreed it should be good to merge as the plan is still to include Hibernate Search in WildFly and EAP.

**Pull request merged**